### PR TITLE
Add duplicate line item(s) action

### LINKED ISSUE #137: Add duplicate line items action

### PR DESCRIPTION
**Pull request by thu5**
This PR adds the ability to duplicate line items. Resolves #137.

### CODE REVIEW

File: server/lib/service/dv360.dart [hunk 1]
```diff
@@ -31,11 +34,58 @@ class DisplayVideo360Client {
...
   /// Runs the rule to manipulate DV360 line items and logs the result.
   Future<void> run(Rule rule, String userId, String ruleId) async {
     for (final target in rule.scope.targets) {
       try {
-        await rule.action.run(this, target);
+        switch (rule.action.runtimeType) {
+          case ChangeLineItemStatusAction:
+            await runStatusAction(rule.action, target);
+            break;
+          case DuplicateLineItemAction:
```
Comment (cathyang-work) [inline server/lib/service/dv360.dart]: Make sure after you merge this pr, the big strategy action is still there.

**Review by cathyang-work — APPROVED** [review]

File: server/lib/service/dv360.dart [hunk 2]
```diff
@@ -31,11 +34,58 @@ class DisplayVideo360Client {
         updateMask: 'entityStatus');
   }
 
+  /// Duplicates the line item to [insertionOrderIdDestination].
+  ///
+  /// The [advertiserIdDestination] is the ID of the destination advertiser.
+  ///
+  /// Throws an [ApiRequestError] if API returns an error.
+  Future<void> duplicateLineItem(Int64 advertiserId, Int64 lineItemId,
```
Comment (gg-qinp) [inline server/lib/service/dv360.dart:44]: fyi: a complete line item contains targeting which is a major part. but this requires much extra effort. so we are good now.

File: server/lib/service/dv360.dart [hunk 3]
```diff
@@ -31,11 +34,58 @@ class DisplayVideo360Client {
...
+    source
+      ..insertionOrderId = insertionOrderIdDestination.toString()
+      ..entityStatus = 'ENTITY_STATUS_DRAFT'
+      ..advertiserId = null
+      ..campaignId = null
+      ..lineItemId = null
+      ..updateTime = null
+      ..partnerCosts = null
+      ..name = null
+      ..flight.dateRange = null
```
Comment (gg-qinp) [inline server/lib/service/dv360.dart]: why not reuse the flight from source line item?

**Review by gg-qinp — APPROVED** [review]

File: server/lib/service/dv360.dart (hunk 3 above)
Comment (thu5) [reply server/lib/service/dv360.dart]: When I was writing it, I wasn't sure if the flight date should be inherited from the destination parent insertion order. When it does inherit, the dateRange is output only (which is why I set it to null).

So we should make the new flightDateType LINE_ITEM_FLIGHT_DATE_TYPE_CUSTOM and have the same dateRange as the source?

Comment (gg-qinp) [reply server/lib/service/dv360.dart]: i see. actually although it is input only, you can still set it, the server just ignore the input, :p, the better way is to add check `If isInherited`, make dateRange = null.

Comment (thu5) [reply server/lib/service/dv360.dart]: Ok gotcha!

**Pull request merged**